### PR TITLE
fix(world): Compute border requests from rasterizers and `@Updates` annotations.

### PR DESCRIPTION
**Pull request by 4Denthusiast**
The previous algorithm for calculating border sizes didn't entirely make sense and couldn't account for borders requested in `@Updates` annotations properly, and entirely ignored borders required by rasterizers, leading to occasional errors like cut-off trees.

### CODE REVIEW

**Comment by pollend** [comment]
seems pretty good to me. I don't see anything wrong with this PR.

**Comment by pollend** [comment]
I tested this with core and I couldn't find any cut trees. any modules in particular that this should be tested against?

**Comment by 4Denthusiast** [comment]
> I tested this with core and I couldn't find any cut trees. any modules in particular that this should be tested against?

Errors with trees occasionally showed up with Caves. The error occurs much more frequently with https://github.com/Terasology/CoreWorlds/pull/18. This is sort of difficult to test unless you already have a world where the error occurs, because it's usually pretty rare anyway.

**Comment by skaldarnar** [comment]
I just checked out this PR  and the one in CoreWorlds, and wanted to start a game of JoshariasSurvival - however, world generation stalls at "Awaiting Character Spawn" and I see this error in the logs:
```
21:48:23.797 [main] ERROR o.t.e.event.internal.EventSystemImpl - Failed to invoke event
java.lang.IllegalArgumentException: Out of bounds: (-24, 41, 34) for region (Min: (-33, -17, -33), Size: (67, 89, 67))
	at org.terasology.world.generation.facets.base.BaseFacet3D.getWorldIndex(BaseFacet3D.java:53)
	at org.terasology.world.generation.facets.base.BaseFieldFacet3D.getWorld(BaseFieldFacet3D.java:47)
	at org.terasology.world.generation.facets.base.BaseFieldFacet3D.getWorld(BaseFieldFacet3D.java:52)
	at org.terasology.caves.CaveToSurfaceProvider.process(CaveToSurfaceProvider.java:105)
	at org.terasology.world.generation.RegionImpl.lambda$getFacet$1(RegionImpl.java:49)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.terasology.world.generation.RegionImpl.getFacet(RegionImpl.java:48)
	at org.terasology.logic.spawner.AbstractSpawner.findSpawnPosition(AbstractSpawner.java:60)
	at org.terasology.logic.spawner.FixedSpawner.getSpawnPosition(FixedSpawner.java:49)
	at org.terasology.core.world.generator.worldGenerators.PerlinFacetedWorldGenerator.getSpawnPosition(PerlinFacetedWorldGenerator.java:53)
	at org.terasology.logic.players.PlayerSystem.onConnect(PlayerSystem.java:156)
	at org.terasology.logic.players.PlayerSystemMethodAccess.invoke(Unknown Source)
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:531)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendStandardEvent(EventSystemImpl.java:292)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:283)
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:205)
	at org.terasology.network.internal.NetworkSystemImpl.connectClient(NetworkSystemImpl.java:882)
	at org.terasology.network.internal.NetworkSystemImpl.joinLocal(NetworkSystemImpl.java:326)
	at org.terasology.engine.modes.loadProcesses.SetupLocalPlayer.step(SetupLocalPlayer.java:45)
	at org.terasology.engine.modes.StateLoading.update(StateLoading.java:237)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:470)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:433)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:409)
	at org.terasology.engine.Terasology.main(Terasology.java:174)
```

**Comment by 4Denthusiast** [comment]
> I just checked out this PR and the one in CoreWorlds, and wanted to start a game of JoshariasSurvival - however, world generation stalls at "Awaiting Character Spawn" and I see this error in the logs. Can this be related?
> 
> **Edit:** this also happens with just the Core gameplay.

This is fixed with https://github.com/Terasology/Caves/pull/12.

**Review by jdrueckert — COMMENTED** [review]
@skaldarnar and @pollend, do you think we should align the methods and naming here with what we did for `BlockRegion`?

File: engine/src/main/java/org/terasology/world/generation/Border3D.java [hunk 1]
```diff
@@ -113,19 +117,30 @@ public Border3D extendBy(int topExtension, int bottomExtension, int sidesExtensi
...
-     * @return The resulting Border3D.
+     * @param other The top value to compare with the instance's top value.
+     * @return The resulting new Border3D.
      */
-    public Border3D maxWith(int topValue, int bottomValue, int sidesValue) {
-        return new Border3D(
-                Math.max(top, topValue),
-                Math.max(bottom, bottomValue),
-                Math.max(sides, sidesValue));
+    public Border3D maxWith(Border3D other) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/Border3D.java:135]: would this be `union` to keep the terminology similar to `BlockRegion`?

File: engine/src/main/java/org/terasology/world/generation/Border3D.java [hunk 2]
```diff
@@ -113,19 +117,30 @@ public Border3D extendBy(int topExtension, int bottomExtension, int sidesExtensi
         return new Border3D(top + topExtension, bottom + bottomExtension, sides + sidesExtension);
     }
 
+    /**
+     * Extends the border by using the sizes of another border.
+     * @param other The border to add to this one
+     * @return The new border with the extra extensions.
+     */
+    public Border3D extendBy(Border3D other) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/world/generation/Border3D.java:125]: Maybe use `expandBy` to be a bit closer to `BlockRegion`?

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**